README.md
# f2u

718525ae-a057-11eb-a679-ebc1ee2a486b Sun, 18 Apr 2021 12:04:59 -0300

# *"To master riding bicycles you have do ride bicycles"*

This is a evolution from what I learning with u2forth, ATMEGA8, gcc assembler and forth implementations as 328eforth, eforth, amforth, avr-forth, gforth, flashforth, punyforth, etc 

Some special insights from jonasforth (https://rwmj.wordpress.com/2010/08/07/jonesforth-git-repository/ https://github.com/nornagon/jonesforth), sectorforth (https://github.com/cesarblum/sectorforth), and lbforth (https://github.com/larsbrinkhoff/lbForth/)

## This --Readme.md-- still is a "brainstorm", really is a documentation of changes 

**"this is a work in progress, not completed"**

(take a look at notes and changes)

( need to learn how to use and format with .md)

https://github.com/matiassingers/awesome-readme

https://silentlad.com/how-to-write-beautiful-and-meaningful-readme.md

# Introduction

*In the chat session Chuck was asked, "How did you come to the conclusion that Forth was too complex, and that sourceless programming was your next move?" His reply was, "Maybe by reading the Forth Standard." [Moore, 2002] <http://www.ultratechnology.com/levels.htm>*

Forth is language based in thread code, with a dictionary of words as named routines and two stacks for arguments in a forever read–eval–print-loop.

The dictionary is a linked list of words, each with name and code. It have two types of words, those called primitives, ad natives ad leaves, which are coded in specific CPU or MCU instructions, and those called compounds, ad twigs, which are sequences with references to words.
 
*"A most important aspect of FORTH is its ability to define new words. New definitions and code are devised continously. Likewise, new constants or variables are created.", Rather, Moore, Hollis, https://library.nrao.edu/public/memos/comp/CDIR_17.pdf*

I want a forth with:

- a minimal inner interpreter and primitives words (clock, uart, interrupts, stacks, math, moves) dependent of a MCU family;

- all compound words independent of any specific MCU family,  without any assembler specifc code inline, like a imutable list with rellocable references.

# Size or Speed ?

*"Keep it Simple"*

Most of Forth implementations goes "runnig for speed" for timming applications or simply to be "the most faster than", but when memory space is the critical limit most of design decisions must take another path.

My choice for design is a Atmega8, a MCU with harvard memory architeture, 4k words (16-bits) program flash memory, 1k bytes (8-bits) static ram memory, 512 bytes of EEPROM,  memory-mapped I/O, one UART, one SPI, one I2C, with 32 (R0 to R31) 8 bits registers, with some that could be used as eight (R16 to R31) 16 bits registers.  A challenge.

There are many low cost MCU with far more resources and pleny of SRAM and flash. Why use an old MCU for hosting Forth ?

Most to refine language paradigms and understood manage memory, RAM and FLASH, and how forth works inside, looking from behind the stacks.

Many challenges to resolve, how make a minimal bios, what basic word must set, how update flash memory, how access internal resources, etc. Learn from previous many implementations of Forth and adapt to survive.

For comparation, in 1979, the PDP-11, Programmed Data Processor, was also eight 16-bit registers, (including one stack pointer R6, and one program counter R7), unifed memory addressing and memory mapped devices. <http://bitsavers.trailing-edge.com/pdf/dec/pdp11/handbooks/PDP11_Handbook1979.pdf>.

And PDP-11 have a successful implementation of Forth in 1974, https://library.nrao.edu/public/memos/comp/CDIR_17.pdf, and of Fig-Forth in 1980, http://www.forth.org/fig-forth/fig-forth_PDP-11.pdf and http://www.stackosaurus.com/figforth.html.

### ; details

Not using of SP intructions (pop, push, call, ret), leaving those for external extensions and libraries;

Only using IJMP to primitives else use indirect push and pull for references address;
     
All references are done using of indirect LD and ST with Z, Y, X, 16 bits registers;
     
All primitive words finish with a rjmp _link, so the (inner + primitives) must be less than +2k words;
      
Uses address pointer is Z (r31:r30) for lpm/spm (flash), lds (sram), sts (sram) instructions;
      
The return stack pointer is Y (r29:r28) for forth return stack;
      
The parameter stack pointer is X (r27:r26) for forth data stack;
      
A working register is W (r25:r24) for forth as acumulator register;
      
A temporary register T (r23:r22);
      
A temporary register N (r21:r20);
      
A instruction register IP (r19:r18);

For convenience

 - reserved r0:r1, used as generic scratch for internal routines (mul, div, etc)
        
 - reserved r2 as generic _work_ 
 
 - reserved r3 as always _zero_, by interrupt ;)
 
 - registers r4 reserver for keep _SREG_, inside interrupts
 
 - register r5 reserved for adjustable prescaler of timer0
 
 - registers r6:r7 reserved for keep 1ms clock ticks counter
 
 - registers r8::r17 are free

# Memory
flash memory from $000 to $FFF ($0000 to $1FFF bytes), splits in read-write flash RWW($000 to $BFF) and no-read-rwrite- flash NRWW ($C00 to $FFF).

sram memory from $0C0 to $45F (1024 bytes)

# Specifics

 For ATmega8 MCU specifics plan, using

   - a MiniCore and optboot;
   - a internal clock of 8MHz and 
   - a uart at 9600, 8N1, asynchronous;
   - a timer at 1ms with 16 bits counter  ~ 65 s;
   - a watch dog at ~ 2.0 s;
   - a pseudo 16bit random generator; 
   - a adapted djb hash generator for 16bits;
   - all 8bits and 16bits math from AVR200 manual;
    
> need explain update/flush flash memory using a sram buffer

# Specifics

  - all dictionary in flash;
  - all constants in flash;
  - all values and variables in sram;
  - eeprom preserves constants;
  - a cell is 16 bits;
  - little endian, low byte at low address;
  - a char is ASCII 7 bits, one byte at SRAM, one cell at stacks.
  - maximum word lenght is 15;
  - four bits flags (IMMEDIATE, COMPILE, HIDEN, TOGGLE) per word;
  - word names are padded with space (0x20)
  - numbers are signed two-complement;
  - all stacks are 18 words (cells);
  - terminal input buffer is 72 bytes, scratch-pad is 64 bytes, pic is 16 bytes;
  - all buffers ends in \0
  - still interrupts can't be nested, SREG breaks.
  
# Notes

- primitive (Leaf) routine does not do any calls. Compound (Twig) routines could do.
- index routines counts downwards until 0, ever, exact as C: for (n=NNNN; n != 0 ; n--) { ~~~ }
- no bounds check, none.
-compare bytes: COMPARE return FALSE or TRUE, only;
- move bytes: only MOVE done, (still no CMOVE upwards, no CMOVE> downwards);
- word names lenght can be 1 to 15, padded with space (0x20);

# Bootstrap fuses

from https://www.engbedded.com/conffuse/

BOOTRST, Flash boot size=256 words Boot address= $0F00
EESAVE, preserve eeprom
SPIEN, serial prrogram downloading SPI enable
SUT_CKSEL, Int Osc, 8MHz; startup time 6CK + 64ms
BODEN, Brown-out detector VCC=2,7V

high 0xD4
low  0xA4 (0xA0 for external clock, 16Mhz)

AVRDUDE:
-U hfuse:w:0xd4:m -U lfuse:w:0xa4:m



changes.md


  10/06/2021, Still no operational

    The basic functional words are done, as minimal set for parse, find, evaluate, number, create, does;
    
    Controls
    a)  testing top of stack:
      
      *standart implementation for*
      (if ... else ... then ...), forward jumps, (1)
      (begin ... again), inconditional backward jump, (2)
      (begin ... until), conditional backward jump, (2)
      
      *non-standart implementation for*
      (while ... repeat), conditional forward jump and inconditional backward jump, (3)
      (while ... again), maybe ?
      att: *repeat jumps to while*, both independent of begin.
    
    b) testing values on return stack *not for now* 
      why not variables I, J, K ? I dont like using return stack for loops.
      (do ... leave ... loop), counter backward jump, test at loop
      (for ... next), counter backward jump, test at for

  12/05/2021,  Still no operational
  
    Start basic assembler for parse, find, evaluate, number, create, does;
    Defined specific functions for flash memory. All changes at dictionary are done in a buffer sram. 
    Created routines, a *flash* for init the buffer, from flash page of HERE, and a *flush* for copy to flash update
  
  08/04/2021, resumes from 2020
  
    The inner interpreter is done and is very small and effcient;
  
    The primitive words are done, as minimal set from forth plus some extras;
  
    But I'm at easter egg of forth:
      I have sources of words as ": word ~~~ ;" and I need a forth done to compile or
      I have sources of words compiled with some forth and need use same forth engine;
  
    Then sectorforth (https://github.com/cesarblum/sectorforth) comes to simplifly all, and I restart again.
  
    The optiboot v8.0 could do program flash memory, as do_spm, so will use it as boot loader.
  

f2u2.S
/*
 *  DISCLAIMER"
 *
 *  Copyright © 2020, Alvaro Barcellos,
 *  Permission is hereby granted, free of charge, to any person obtaining
 *  a copy of this software and associated documentation files (the
 *  "Software"), to deal in the Software without restriction, including
 *  without limitation the rights to use, copy, modify, merge, publish,
 *  distribute, sublicense, and/or sell copies of the Software, and to
 *  permit persons to whom the Software is furnished to do so, subject to
 *  the following conditions"
 *
 *  The above copyright notice and this permission notice shall be
 *  included in all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 *
*/

/*

  inspired on eforth from books and papers from C. H. Ting
 
  just experience for small microcontrolers
 
  emulates a virtual uCi,  
	
  still no interrupts
 
  still no multitasks
	
	the little endian method, the low byte of the address or value occupies the lower memory address. 

	Notation: top is at right, eg ROT (w1 w2 -- w1 w2 w1) w1 is 2nd at stack, w2 is 1st at stack
 
;---------------------------------------------------------------------
; dictionary structure
;
 .word  link    link to previous
 .byte  size    sizeof name OR mask 0x1F
 .ascii "name" name, plus a pad byte with BL (0x32) if size is odd
 .word  param,  many

;----------------------------------------------------------------------

18/04/2021

	Rewrite all from u2f*, using a new minimal forth engine as LAST word 

*/

    .NOLIST

    .LIST

;----------------------------------------------------------------------
; get some magic constants
 
#include <avr/io.h>

;----------------------------------------------------------------------

#include "equates.inc" 

;----------------------------------------------------------------------

#include "macros.inc"

;----------------------------------------------------------------------

#include "sram.inc"

;----------------------------------------------------------------------

#include "boot.inc"

;----------------------------------------------------------------------
; start of dictionary
;
;----------------------------------------------------------------------

;----------------------------------------------------------------------
;
; any leaf must start with NOOP and end with POON
;
;----------------------------------------------------------------------
;
; it is the forth engine,
;
; any twig must end with a reference to LAST
;
; any leaf must start with NOOP and ends with POON
;
; zzz size must be mask or 0xE0
;
; instruction pointer r31:r30 as Z 
; return stack pointer r29:r28 as Y
; working register r25:r24 as W
; scratch register r0
;
; only LPM access flash ? YES!
;
;	if a vocabulary is in ram ? for atmega8 it will could not.
;
; all forth constants in flash
;
; all forth variables in sdram
;
;----------------------------------------------------------------------
.section .text
.org 0x100
.global forth
forth:
;----------------------------------------------------------------------
; inner interpreter, must have another name no LAST
;
HEADER "_LAST", "_LAST"
; does nothing and mark instructions code
	NOOP

; pull ip from rsp
_EXIT: 			
	rspull ips_low, ips_high

; load w with contents of cell at ip, only works in program memory (flash)  
_NEXT:			
	lpm wrk_low, Z+
	lpm wrk_high, Z+

; if zero then is a exec 
	cp wrk_low, wrk_high
	brbs BIT_ZERO, _EXEC

; else is a reference
_ENTER:			
	rspush ips_low, ips_high

; push ip into rsp
	movw ips_low, wrk_low
	rjmp _NEXT

_EXEC:
	ijmp

;----------------------------------------------------------------------
; 
; using a classic, tos, (1st, 2nd, 3rd, 4th, ...)
;

;----------------------------------------------------------------------
; ok ( w -- ) ( -- W) 
; 	F18A push
HEADER ">R", "TOR"
	NOOP
	pspull wrk_low, wrk_high
	rspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( -- w ) ( w -- )
;	F18A pull
HEADER "R>", "RTO"
	NOOP
	rspull wrk_low, wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( -- w ) ( w -- w )
HEADER "R@", "RAT"
	NOOP
	rspull wrk_low, wrk_high
	rspush wrk_low, wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( -- rsp)  
HEADER "RP@", "RPAT"
	NOOP
	movw wrk_low, rsp_low
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( rsp -- )
HEADER "RP!", "RPTO"
	NOOP
	pspull wrk_low, wrk_high
	movw rsp_low, wrk_low
	POON

;----------------------------------------------------------------------
; ok ( -- psp )
HEADER "SP@", "SPAT"
	NOOP
	movw wrk_low, psp_low
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( psp -- )
HEADER "SP!", "SPTO"
	NOOP
	pspull wrk_low, wrk_high
	movw psp_low, wrk_low
	POON

;----------------------------------------------------------------------
; ok ( -- w )
;	F18A
HEADER "PUSH", "PUSH"
	NOOP
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( w -- )
;	F18A drop
HEADER "DROP", "DROP"
PULL:
	NOOP
	pspull wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( w -- w w)
; 	F18A dup
HEADER "DUP", "DUP"
	NOOP
	pspull wrk_low, wrk_high
	pspush wrk_low, wrk_high
	pspush wrk_low, wrk_high
	POON 

;----------------------------------------------------------------------
; ok ( w1 w2 -- w2 w1 )
HEADER "SWAP", "SWAP"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	pspush wrk_low, wrk_high
	pspush tos_low, tos_high
	POON

;----------------------------------------------------------------------
; ok ( w1 w2 -- w1 w2 w1 )
; 	F18A over
HEADER "OVER", "OVER"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	pspush tos_low, tos_high
	pspush wrk_low, wrk_high
	pspush tos_low, tos_high
	POON

;----------------------------------------------------------------------
; ok ( w1 w2 w3 -- w2 w3 w1 )
HEADER "ROT", "ROT"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	pspull nds_low, nds_high
	pspush wrk_low, wrk_high
	pspush nds_low, nds_high
	pspush tos_low, tos_high
	POON

;----------------------------------------------------------------------
; ok ( w -- 0 | w w )
HEADER "?DUP", "ZDUP"
	NOOP
	pspull wrk_low, wrk_high
	pspush wrk_low, wrk_high
	cp wrk_low, wrk_high
	brbs BIT_ZERO, NZDUP_1
	pspush wrk_low, wrk_high
NZDUP_1:
	POON 

;----------------------------------------------------------------------
;ok ( u -- )
;	F18A	T=0
HEADER "?BRANCH", "QBRANCH"
	NOOP
	pspull wrk_low, wrk_high
	cp wrk_low, wrk_high
	brbs BIT_ZERO, BRANCH_END
_ADVANCE:
	rspull wrk_low, wrk_high
	adiw wrk_low, CELL_SIZE
	rspush wrk_low, wrk_high
_BRANCH_END:
	POON

;----------------------------------------------------------------------
;ok ( branch next reference )
HEADER "BRANCH", "BRANCH"
	NOOP
	POON

;----------------------------------------------------------------------
;ok ( branch next reference )
HEADER "EXEC", "EXEC"
	NOOP
	pspull wrk_low, wrk_high
	rspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( w1 w2 -- w2+w1 )
HEADER "+", "PLUS"
	NOOP
	clc
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	add wrk_low, tos_low
	adc wrk_high, tos_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( w1 w2 -- w2-w1 )
HEADER "-", "MINUS" 
	NOOP
	clc
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	sub wrk_low, tos_low
	sbc wrk_high, tos_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ( w1 w2 -- 0 | -1)
HEADER "=", "EQUAL"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	cp wrk_low, tos_low
	cpc wrk_high, tos_high
	brne FFALSE+2
	rjmp FTRUE+2

;----------------------------------------------------------------------
; nok ( w1 w2 -- 0 | -1 )
HEADER "<", "LTHAN"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	cp wrk_low, tos_low
	cpc wrk_high, tos_high
	brcs FFALSE+2
	rjmp FTRUE+2

;----------------------------------------------------------------------
; nok ( w1 w2 -- 0 | -1 )
HEADER ">", "GTHAN"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	cp wrk_low, tos_low
	cpc wrk_high, tos_high
	brcc FFALSE+2
	rjmp FTRUE+2

;----------------------------------------------------------------------
; ok ( -- -1 )
HEADER "TRUE", "FTRUE"
	NOOP
	ser wrk_low
	ser wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( -- 0 )
HEADER "FALSE", "FFALSE"
	NOOP
	clr wrk_low
	clr wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w1 w2 -- w2 AND w1 )
HEADER "AND", "AND"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	and  wrk_high, tos_high
	and  wrk_low, tos_low
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w1 w2 -- w2 OR w1 )
HEADER "OR", "OR"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	or   wrk_high, tos_high
	or   wrk_low, tos_low
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w1 w2 -- w2 XOR w1 )
HEADER "XOR", "XOR"
	NOOP
	pspull wrk_low, wrk_high
	pspull tos_low, tos_high
	eor   wrk_high, tos_high
	eor   wrk_low, tos_low
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w -- w * 2 ) same as shift left << 1, unsigned
HEADER "2*", "SHL"
	NOOP
	pspull wrk_low, wrk_high
	clc
	lsl   wrk_low
	rol   wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w -- w / 2 ) same as shift right >> 1, unsigned
HEADER "2/", "SHR"
	NOOP
	pspull wrk_low, wrk_high
	clc
	lsr   wrk_low
	ror   wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w -- w * 2 ) same as shift left << 1, signed
HEADER "2*A", "SHLA"
	NOOP
	pspull wrk_low, wrk_high
	clc
	bst wrk_high, 7
	cbr	wrk_high, 7
	lsl   wrk_low
	rol   wrk_high
	bld wrk_high, 7
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( w -- w / 2 ) same as shift right >> 1, signed
HEADER "2/A", "SHRA"
	NOOP
	pspull wrk_low, wrk_high
	clc
	bst wrk_high, 7
	cbr	wrk_high, 7
	lsr   wrk_low
	ror   wrk_high
	bld wrk_high, 7
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;ok ( nAnB -- nBnA ) swap bytes in a word
HEADER "><", "ENDIAN"
	NOOP
	pspull wrk_low, wrk_high
	pspush wrk_high, wrk_low
	POON

;----------------------------------------------------------------------
;  os doLIT ( -- w ), from flash program memory to data stack
HEADER "DOLIT", "DOLIT"
	NOOP
	rspull ips_low, ips_high
; load from flash program memory
	pmload wrk_low, wrk_high
	rspush ips_low, ips_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;  os doVAR ( -- w ) from sram memory to data stack
HEADER "DOVAR", "DOVAR"
	NOOP
	rspull ips_low, ips_high
; load from static ram memory
	srload wrk_low
	srload wrk_high
	rspush ips_low, ips_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;  os doCON ( -- w ) from sram memory to data stack
HEADER "DOCON", "DOCON"
	NOOP
	rspull ips_low, ips_high
; load from static ram memory
	srload wrk_low
	srload wrk_high
	rspush ips_low, ips_high
	movw ips_low, wrk_low
; load from static ram memory
	srload wrk_low
	srload wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
; ok ! to (x a -- )
HEADER "!", "TO"
	NOOP
	pspull ips_low, ips_high
	pspull wrk_low, wrk_high
	srsave wrk_low
	srsave wrk_high
	POON

;----------------------------------------------------------------------
; ok C! ( c a -- )
HEADER "C!", "CTO"
	NOOP
	pspull ips_low, ips_high
	pspull wrk_low, wrk_high
	srsave wrk_low
	POON

;----------------------------------------------------------------------
; ok @ at ( a -- w )
HEADER "@", "AT"
	NOOP
	pspull ips_low, ips_high
    srload wrk_low
    srload wrk_high
	pspush wrk_low, wrk_high
    POON

;----------------------------------------------------------------------
; ok C@ cat
HEADER "C@", "CAT"
	NOOP
	pspull ips_low, ips_high
    srload wrk_low
	clr wrk_high
	pspush wrk_low, wrk_high
    POON

;----------------------------------------------------------------------
;  ok 0>	must be signed
HEADER "0>", "ZEROGT"
	NOOP
	pspull wrk_low, wrk_high
    sbiw wrk_low, 1
    brge _ZEROGT1
    rjmp FTRUE
_ZEROGT1:
    rjmp FFALSE

;----------------------------------------------------------------------
;  ok 0<
HEADER "0<", "ZEROLT"
	NOOP
	pspull wrk_low, wrk_high
    sbiw wrk_low, 1
    brge _ZEROLT1
    rjmp FFALSE
_ZEROLT1:
    rjmp FTRUE

;----------------------------------------------------------------------
; ok 0=
HEADER "0=", "ZEROEQ"
	NOOP
	pspull wrk_low, wrk_high
	cp wrk_high, wrk_low
	brbs BIT_ZERO, _ZEROEQ1
    rjmp FFALSE
_ZEROEQ1:
    rjmp FTRUE

;----------------------------------------------------------------------
; ok INV
HEADER "INV", "INV"
	NOOP
	pspull wrk_low, wrk_high
    com wrk_low
    com wrk_high
	pspush wrk_low, wrk_high
    POON

;----------------------------------------------------------------------
; ok NEG
HEADER "NEG", "NEG"
	NOOP
	pspull wrk_low, wrk_high
    neg wrk_low
    neg wrk_high
	pspush wrk_low, wrk_high
    POON

;----------------------------------------------------------------------
; ok ABS
HEADER "ABS", "ABS"
	NOOP
	pspull wrk_low, wrk_high
    cbr wrk_high, 7
	pspush wrk_low, wrk_high
    POON

;----------------------------------------------------------------------
;
; compare bytes from source++ to destination++, decrease count--
; ( a1 u1 a2 u2 -- flag)
;
HEADER "COMPARE", "COMPARE"
	NOOP
    ; WARNING No bound check and counter equal 0 moves nothing

	; pull lenght 
	pspull tos_low, tos_high

	; pull address
	pspull ips_low, ips_high
	
	; pull lenght
	pspull nds_low, nds_high

    ; pull address
	pspull wrk_low, wrk_high

	; push psp
	rspush psp_low, psp_high

	; copy address
	movw psp_low, wrk_low

    ; no offset

_COMPARE_LOOP:

	cp  tos_low, tos_high
	brbs BIT_ZERO, _COMPARE_EQUAL

	cp  nds_low, nds_high
	brbs BIT_ZERO, _COMPARE_EQUAL

	ld wrk_low, X+
	ld wrk_high, Z+
	cp wrk_low, wrk_high
	brbc BIT_ZERO, _COMPARE_DIFF

	subi tos_low, 1
	sbci tos_high, 0

	subi nds_low, 1
	sbci nds_high, 0

    ; again
	rjmp _COMPARE_LOOP

_COMPARE_EQUAL:
	
	cp tos_low, nds_low
	brbc BIT_ZERO, _COMPARE_DIFF

	cp tos_high, nds_high
	brbc BIT_ZERO, _COMPARE_DIFF

	clr wrk_low
	clr wrk_high
	rjmp _COMPARE_ENDS

_COMPARE_DIFF:
	ser wrk_low
	ser wrk_high

_COMPARE_ENDS:
	rspull psp_low, psp_high
	pspush wrk_low, wrk_high	
	POON

;----------------------------------------------------------------------
;
; CMOVE copy bytes from source++ to destination++, decrease count--
HEADER "CMOVE", "MOVE"
	NOOP
    ; WARNING No bound check and counter equal 0 moves nothing

	; pull origin
	pspull tos_low, tos_high
	
	; pull destination
	pspull ips_low, ips_high

    ; pull counter
	pspull wrk_low, wrk_high

	; test if zero
	cp wrk_low, wrk_high
	brbs BIT_ZERO, _CMOVE_ENDS

	; push psp
	rspush psp_low, psp_high

	; copy orign
	movw psp_low, tos_low

    ; no offset

_CMOVE_LOOP:

    ; move a byte
    ld _work_, X+
    st Z+, _work_
    
    ; decrease counter
	subi wrk_low, 1
	sbci wrk_high, 0

    ; again
	brbc BIT_ZERO, _CMOVE_LOOP

_CMOVE_ENDS:
	; pull psp
	rspull psp_low, psp_high
	POON

;----------------------------------------------------------------------
;
; CMOVE> copy bytes from source+count-- to destination--, decrease count--
HEADER "CMOVE>", "CMOVEGT"
	NOOP
    ; WARNING No bound check and counter equal 0 moves nothing

	; pull origin
	pspull tos_low, tos_high
	
	; pull destination
	pspull ips_low, ips_high

    ; pull counter
	pspull wrk_low, wrk_high

	; test if zero
	cp wrk_low, wrk_high
	brbs BIT_ZERO, _CMOVE_ENDS

	; push psp
	rspush psp_low, psp_high

	; copy orign
	movw psp_low, tos_low

    ; do offset
	add ips_low, wrk_low
	adc ips_high, wrk_high
	add psp_low, wrk_low
	adc psp_high, wrk_high

_CMOVEGT_LOOP:
    ; move a byte
    ld _work_, -X
    st -Z, _work_
    
    ; decrease counter
	subi wrk_low, 1
	sbci wrk_high, 0
	brbc BIT_ZERO, _CMOVEGT_LOOP

_CMOVEGT_ENDS:
	; pull psp
	rspull psp_low, psp_high
	POON

;----------------------------------------------------------------------
;  data space pointer align to CELL, atmega is always
HEADER "ALIGN", "ALLIGN"
	NOOP
	pspull wrk_low, wrk_high
	adiw wrk_low, 1
	and wrk_low, 0xFE
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok CELL is 2 ; size of a cell
;
HEADER "CELL", "CELLSZ"
	NOOP
    ldi wrk_low, CELL_SIZE
    clr wrk_high
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok CELL is 2, multiply by 2, easy way
;
HEADER "CELLS", "CELLS"
	NOOP
    rjmp SHL+2

;----------------------------------------------------------------------
;
; ok 
;
HEADER "CELL+", "CELLPLUS"
	NOOP
	pspull wrk_low, wrk_high
	adiw wrk_low, CELL_SIZE
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "CELL-", "CELLMINUS"
	NOOP
	pspull wrk_low, wrk_high
	sbiw wrk_low, CELL_SIZE
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "1+", "ONEPLUS"
	NOOP
	pspull wrk_low, wrk_high
	adiw wrk_low, 1
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "1-", "ONEMINUS"
	NOOP
	pspull wrk_low, wrk_high
	sbiw wrk_low, 1
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "2+", "TWOPLUS"
	NOOP
	pspull wrk_low, wrk_high
	adiw wrk_low, 2
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "2-", "TWOMINUS"
	NOOP
	pspull wrk_low, wrk_high
	sbiw wrk_low, 2
	pspush wrk_low, wrk_high
	POON

;----------------------------------------------------------------------
;
; ok 
;
HEADER "+!", "PLUSTO"
	NOOP
	pspull ips_low, ips_high
	pspull wrk_low, wrk_high
	srsave wrk_low
	srsave wrk_high
	POON


;----------------------------------------------------------------------
; fake
HEADER  "WARM", "WARM"
	NOOP
	POON

;----------------------------------------------------------------------
;
#include "variables.inc"

#include "constants.inc"

#include "cold.inc"

#include "uart.inc"

#include "hash.inc"

#include "rand.inc"

#include "longs.inc"

;#include "twigs.inc"

#include "efesp32.inc"
